libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,732 @@
+{
+    AP_SerialManager& serial_manager = AP::serialmanager();
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FETtecOneWire, 0);
+    if (_uart) {
+        _uart->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        _uart->set_unbuffered_writes(true);
+        _uart->set_blocking_writes(false);
+        _uart->begin(500000);
+    }
+}
+
+
+void AP_FETtecOneWire::update()
+{
+
+    if (!_initialised) {
+        _initialised = true;
+        init();
+        _last_send_us = AP_HAL::micros();
+        return;
+    }
+
+    if (_uart == nullptr) {
+        return;
+    }
+
+    if (!_uart->is_dma_enabled()) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "FTW UART needs DMA");
+    }
+    // tell SRV_Channels about ESC capabilities
+    const uint16_t mask = uint16_t(motor_mask.get());
+    SRV_Channels::set_digital_outputs(mask, 0);
+
+    // get ESC set points, stop as soon as there is a gap
+    uint16_t motor_pwm[MOTOR_COUNT_MAX] = {1000};
+
+    for (uint8_t i = 0; i < MOTOR_COUNT_MAX; i++) {
+        SRV_Channel* c = SRV_Channels::srv_channel(i);
+        if (c == nullptr) {
+            break;
+        }
+        motor_pwm[i] = constrain_int16(c->get_output_pwm(), 0, 2000);
+    }
+
+#if HAL_WITH_ESC_TELEM
+    // TLM recovery, if e.g. a power loss occurred but FC is still powered by USB.
+    if (!AP::arming().is_armed()) {
+        uint8_t nr_escs = 0;
+        while (mask & (1<<nr_escs)) {
+            nr_escs++;
+        }
+
+        const uint8_t num_active_escs = AP::esc_telem().get_num_active_escs();
+        if (num_active_escs < nr_escs) {
+            const uint32_t now = AP_HAL::millis();
+            if ((now-_lastESCScan) > 5000) { //Scan timeout fix here?
+                GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "FTW found only %i of %i ESCs", num_active_escs, nr_escs);
+                _lastESCScan = now;
+                _scan_active = 0;
+                _setup_active = 0;
+                _set_full_telemetry_active = 0;
+            }
+        }
+    }
+
+
+    // receive and decode the telemetry data from one ESC
+    // but do not process it any further to reduce timing jitter in the escs_set_values() function call
+    TelemetryData t {};
+    int16_t centi_erpm;
+    uint16_t tx_err_count;
+    uint8_t tlm_ok = 0; //decode_single_esc_telemetry returns 1 if telemetry is ok, 0 if its waiting and 2 if there is a crc mismatch.
+    uint8_t tlm_from_id = 0;
+    if (_requested_telemetry_from_esc) {
+        tlm_ok = decode_single_esc_telemetry(t, centi_erpm, tx_err_count, tlm_from_id);
+    }
+    if (_requested_telemetry_from_esc == _found_escs_count) { //if found esc number is reached restart request counter
+        _requested_telemetry_from_esc = 1; // restart from the first ESC
+    } else {
+        _requested_telemetry_from_esc++;
+    }
+#endif
+
+
+    uint32_t txSpace =  _uart->txspace(); //Ugly workaround.
+    if (txSpace==511 && !_uart->tx_pending()) { //Ugly workaround.

```

Comment:
what the heck? why 511? seems to be assuming buffer size of 512. We need to know what the actual protocol limitation is here

Reply:
Afaik the uart buffer is fixed at 512 bytes and txspace() returns 511 if its empty. You can see in my pictures from 12 May that the uart buffer is not always send regular on telem2 port. As I like the ESC to have most current values, old values can be discard. But the function (flush) that clears the buffer is (as far as I can see it) only implemented for USB.
So that workaround assures, that only one ESC command is stored in the buffer and as far as the buffer is send the newest comes into the queue after only one old.

Reply:
Not sure about the workaround, but it needs a comment

Reply:
512 is specific to a fc board, other boards will have other values. Maybe Telem2 did not have DMA activated when you tested it. Do we still need this?

Reply:
I will try it with full-duplex on telem1 port without that workaround. But without something like that more than one old package could be buffered and it can be become delayed by a few packages... It took me quiet a while to find something existing and useful without touching to much stuff.
If I understand it correct the AP_HAL::UARTDriver is using ByteBuffer. ByteBuffer offers is_empty() that would be useful and independent of the buffer size. But we would need to add a function to AP_HAL::UARTDriver like txBuffer.is_empty(). Would that be a good way?

I had seen that 512 is hardcoded in libraries/AP_HAL_SITL/UARTDriver.cpp. line 560... I can't find something like get_size(). I do not know why 511 results when the buffer is empty, but the Byte calculation in space() adds a -1 to the calculation. I didn't question that (Line 76 in RingBuffer.cpp).

Reply:
I now removed that code